sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala

```diff
@@ -57,173 +55,170 @@ abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate w
 
   override def dataType: DataType = DoubleType
 
-  override def inputTypes: Seq[AbstractDataType] = Seq(NumericType)
+  override def inputTypes: Seq[AbstractDataType] = Seq(DoubleType)
+
+  protected val count = AttributeReference("count", DoubleType, nullable = false)()
+  protected val avg = AttributeReference("avg", DoubleType, nullable = false)()
+  protected val m2 = AttributeReference("m2", DoubleType, nullable = false)()
+  protected val m3 = AttributeReference("m3", DoubleType, nullable = false)()
+  protected val m4 = AttributeReference("m4", DoubleType, nullable = false)()
+
+  override val aggBufferAttributes = Seq(count, avg, m2, m3, m4).take(momentOrder + 1)
+
+  override val initialValues: Seq[Expression] = Seq(
+    /* count = */ Literal(0.0),
+    /* avg = */ Literal(0.0),
+    /* m2 = */ Literal(0.0),
+    /* m3 = */ Literal(0.0),
+    /* m4 = */ Literal(0.0)
+  ).take(momentOrder + 1)
+
+  override lazy val updateExpressions: Seq[Expression] = {
+    val n = count + Literal(1.0)
+    val delta = child - avg
+    val deltaN = delta / n
+    val newAvg = avg + deltaN
+    val newM2 = m2 + delta * (delta - deltaN)
+
+    val delta2 = delta * delta
+    val deltaN2 = deltaN * deltaN
+    val newM3 = m3 - Literal(3.0) * deltaN * newM2 + delta * (delta2 - deltaN2)
```

Comment:
If my understanding is correct, there exists some unnecessary expression calculation when we don't need 3rd- or 4th-order statistics. But this is a one-time overhead because `updateExpression` is only called once before codegen. If this is correct, it would be helpful to leave a comment here to explain why we don't need `if` branches.